sdks/python/apache_beam/io/gcp/bigtableio_test.py

```diff
@@ -0,0 +1,218 @@
+      for i in range(pos_start, pos_start + row_count):
+        yield PartialRowData(self._key_bytes('beam_key%07d' % i))
+
+    mock_read_rows.return_value = _mock_read_list()
+    bigtable = BigtableSource(self.project_id, self.instance_id, self.table_id)
+    rows = list(bigtable.read(bigtable.get_range_tracker()))
+    self.assertEqual(len(rows), row_count)
+    for row in rows:
+      self.assertIsInstance(row, PartialRowData)
+      self.assertNotEqual(row.row_key, b'')
+
+  @mock.patch.object(Table, 'sample_row_keys')
+  @mock.patch.object(Table, 'read_rows')
+  def test_read_small_table(self, mock_read_rows, mock_sample_row_keys):
+    row_count = 10000
+
+    def _mock_read_rows(): # 12.2 KB
+      for i in range(0, row_count):
+        yield PartialRowData(self._key_bytes('beam_key%07d' % i))
+
+    mock_sample_row_keys.return_value = list(self._mock_sample_keys())
+    mock_read_rows.return_value = _mock_read_rows()
+    source = BigtableSource(self.project_id, self.instance_id, self.table_id)
+
+    for split_bundle in source.split(None):
+      range_tracker = source.get_range_tracker(split_bundle.start_position, split_bundle.stop_position)
+      rows = list(source.read(range_tracker))
+      self.assertEqual(len(rows), row_count)
+
+      for row in rows:
+        self.assertIsInstance(row, PartialRowData)
+
+  def _mocking_read_rows(self, **kwargs):  # 12.2 KB
+    start_key = kwargs['start_key'] if kwargs['start_key'] is not None else b''
+    end_key = kwargs['end_key']
+
+    from apache_beam.io.range_trackers import LexicographicKeyRangeTracker
+
+    index_start = 0
+    index_stop = RANGES_DICT[b''][1]
+    if start_key != b'':
+      fraction = LexicographicKeyRangeTracker.position_to_fraction(start_key)
+      index_start = int(fraction * index_stop)
+    elif end_key is not None:
+      fraction = LexicographicKeyRangeTracker.position_to_fraction(end_key)
+      index_stop = int(fraction * index_stop)
+
+    for i in range(index_start, index_stop):
+      yield PartialRowData(self._key_bytes('{}\x00{:07d}'.format(start_key, i)))
+
+  @mock.patch.object(Table, 'sample_row_keys')
+  @mock.patch.object(Table, 'read_rows')
+  def test_read_table(self, mock_read_rows, mock_sample_row_keys):
+    mock_sample_row_keys.return_value = list(self._mock_sample_keys(KEYS_2))
+    mock_read_rows.side_effect = self._mocking_read_rows
+    source = BigtableSource(self.project_id, self.instance_id, self.table_id)
+
+    # TODO: Need to implement mock-reader and row counter
+
+    bundles = [b for b in source.split(None)]
+    self.assertEqual(len(bundles), len(KEYS_2)) # Create One Row for each Bundle in the read_rows method.
+
+  @mock.patch.object(Table, 'sample_row_keys')
+  @mock.patch.object(Table, 'read_rows')
+  def test_dynamic_work_rebalancing(self, mock_read_rows, mock_sample_row_keys):
+    """ [INTENTIONALLY BYPASSED]
+
+		This test has been temporarily disabled due to the issues with the LexicographicKeyRangeTracker() class
```

Comment:
I think dynamic work reblanancing related issues are very serious and could indicate data loss or data duplication. So I suggest getting to the bottom of the underlying issue here before submitting this PR. Or at least we should find a way to disable dynamic work rebalancing for this source somehow till this is resolved (may involve updating the range tracker to reject split requests).

Reply:
I agree that they are serious.  However, I don't think that our team can solve the problem.  The problem is that `LexicographicKeyRangeTracker` doesn't work well with `source_test_utils.assert_split_at_fraction_exhaustive()`.  

I personally would prefer disabling dynamic work rebalancing, and have the core Beam team work on making `LexicographicKeyRangeTracker` compatible with `source_test_utils.assert_split_at_fraction_exhaustive()`.

Reply:
That makes sense. Can you please file a JIRA with details about the issue ? If you think it's an issue with the test framework (source_test_utils) I can look into that.

I think you should be able to disable dynamic work rebalancing by sub-classing the range tracker and overriding try_split() method to always return 'None'.

Reply:
As the BoundedSource is no longer used, the RangeTracker is no longer necessary and has been removed as well.